aptos-move/aptos-aggregator/src/types.rs

```diff
@@ -23,11 +48,118 @@ impl AggregatorID {
+}
+
+pub(crate) fn string_from_bytes(bytes: Vec<u8>) -> Value {
+    Value::struct_(Struct::pack(vec![Value::vector_u8(bytes)]))
+}
+
+fn bytes_from_string(value: Value) -> PartialVMResult<Vec<u8>> {
+    value
+        .value_as::<Struct>()?
+        .unpack()?
+        .next()
+        .ok_or_else(|| {
+            // This should never happen, but the check is still here
+            // in case layout and value diverge.
+            PartialVMError::new(StatusCode::INDEX_OUT_OF_BOUNDS)
+        })?
+        .value_as::<Vec<u8>>()
+}
+
+impl TryIntoMoveValue for AggregatorID {
+    type Error = PartialVMError;
+
+    fn try_into_move_value(self, layout: &MoveTypeLayout) -> Result<Value, Self::Error> {
+        Ok(match layout {
+            MoveTypeLayout::U64 => Value::u64(self.0),
+            MoveTypeLayout::U128 => Value::u128(self.0 as u128),
+            layout if is_string_layout(layout) => {
+                // WARNING: when changing this mapping, one has to update it
+                // for the reverse case.
+                string_from_bytes(self.0.to_le_bytes().to_vec())
+            },
+            // We use ID to value conversion in deserialization.
+            _ => {
+                return Err(PartialVMError::new(StatusCode::VALUE_DESERIALIZATION_ERROR)
+                    .with_message(format!(
+                        "Failed to convert {:?} into a Move value with {} layout",
+                        self, layout
+                    )))
+            },
+        })
+    }
+}
+
+impl TryFromMoveValue for AggregatorID {
+    type Error = PartialVMError;
+    type Hint = ();
+
+    fn try_from_move_value(
+        layout: &MoveTypeLayout,
+        value: Value,
+        _maybe_hint: Option<&Self::Hint>,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self::new(match layout {
+            MoveTypeLayout::U64 => value.value_as::<u64>()?,
+            // SAFETY: We can downcast because IDs are contructed from
+            //         values <= u64::MAX.
+            MoveTypeLayout::U128 => value.value_as::<u128>()? as u64,
+            layout if is_string_layout(layout) => {
+                let bytes = bytes_from_string(value)?;
+                // SAFETY: We can convert to u64 because IDs which are
+                //         contructed from bytes use `u64::to_le_bytes`.
+                u64::from_le_bytes(
+                    bytes
+                        .as_slice()
+                        .try_into()
+                        .expect("Should be able to construct u64 from bytes"),
+                )
+            },
+            // We use value to ID conversion in serialization.
+            _ => {
+                return Err(PartialVMError::new(StatusCode::VALUE_SERIALIZATION_ERROR)
+                    .with_message(format!(
+                        "Failed to convert a Move value with {} layout into an identifier",
+                        layout
+                    )))
+            },
+        }))
+    }
+}
+
```

Comment:
this all seems similar to the code in aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

maybe worth deduping if possible? (fine if to do in a separate PR)

Reply:
I made it nicer, and refactored some code into common utils. I still think we can do it better, but keeping for now because the rest can come separately